### PR TITLE
change to tinystr for timezone/metazone id identifiers

### PR DESCRIPTION
**Pull request by samchen61661**
Change to tinystr for timezone/metazone id identifiers (https://github.com/unicode-org/icu4x/issues/528)

### CODE REVIEW

File: provider/datagen/src/cldr/transform/time_zones/convert.rs
Comment (samchen61661) [inline provider/datagen/src/cldr/transform/time_zones/convert.rs:163]: It looks like I am not able to use macro `tinystr` here.

File: provider/datagen/src/cldr/transform/time_zones/convert.rs
Comment (samchen61661) [inline provider/datagen/src/cldr/transform/time_zones/convert.rs:208]: I forgot to update here in my previous PRs. And I believe this change updated json files in this PR.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/datetime/src/date.rs
Comment (nordzilla) [reply components/datetime/src/date.rs:12]: I'm happy using the newtype pattern here as well.

**Review by nordzilla — CHANGES_REQUESTED** [review]
This is looking good to me, pending the changes Shane mentioned regarding using `TinyAsciiStr`.

**Comment by dminor** [comment]
@samchen61661 It looks like you need to fix `cargo fmt`.

**Review by nordzilla — CHANGES_REQUESTED** [review]
@samchen61661 

As Dan mentioned, you just have to run `cargo fmt`. 

regarding the `rustc` panic that we were encountering, I am seeing the exact same behavior on the `main` ICU4X branch, and I don't think that your patch is causing this, Sam. 

I think we should be clear to merge this after you run the formatter.

**Review by nordzilla — DISMISSED** [review]
This all looks good to me. 

I made a few minor commits myself to get this merged in time for the release.

Given that I made some edits, I'd appreciate a quick final look @sffc.

File: provider/datagen/src/transform/cldr/cldr_serde/time_zones/bcp47_tzid.rs
Comment (sffc) [inline provider/datagen/src/transform/cldr/cldr_serde/time_zones/bcp47_tzid.rs:33]: Suggestion, here and in meta_zones.rs: Use the upgraded types

**Review by sffc — APPROVED** [review]

File: components/datetime/tests/patterns/structs/time_zones.rs
Comment (nordzilla) [reply components/datetime/tests/patterns/structs/time_zones.rs:12]: I explicitly changed these to be `TinyAsciiStr` because it was causing the tests to fail without `--features serde` enabled.

The alternative would be to make a test-only, always-serde-enabled type to read the data-driven test JSON and then convert that into the real type as we've done before. But I don't see a point because `TinyAsciiStr` essentially is that already.

Comment (sffc) [reply components/datetime/tests/patterns/structs/time_zones.rs:12]: OK fair enough. I approved the PR so feel free to hit the merge button :)

File: components/datetime/src/date.rs
Comment (samchen61661) [reply components/datetime/src/date.rs:82]: created https://github.com/unicode-org/icu4x/issues/1857 to track

**Review by sffc — APPROVED** [review]

**Pull request merged**
